---
title: "phate/slingshot"
author: "Nate Mullin"
date: "7/24/2022"
output: html_document
---

GOALS: Seems like PHATE dim reduction + Slingshot for trajectories should work OKAY. Now going to:
  1) re-do PHATE and try to get a finalized embedding for PRC lineage
  2) do a finalized cluster annotation of PRC states based on embedding
  3) map trajectories with SLINGSHOT.

```{r setup, include=FALSE}
# using https://htmlpreview.github.io/?https%3A%2F%2Fgithub.com%2FKrishnaswamyLab%2FphateR%2Fblob%2Fmaster%2Finst%2Fexamples%2Fbonemarrow_tutorial.html=

library(Seurat)
library(sctransform)
library(tidyverse)
library(RColorBrewer)
library(phateR)
library(ggplot2)
library(readr)
library(viridis)
library(Rmagic)
library(slingshot)
library(cowplot)
library(ggpubr)
library(glmGamPoi)
library(ggrepel)
library(tradeSeq)
library(scales)
library(clusterExperiment)
library(scCustomize)

set.seed(1234)

```

```{r}
 # load in the seurat object with UMAP
load(".../organoid_all_tp.RData")
```



```{r fig.width=10, fig.height=10}
# use this pipeline for taking counts from seurat object, running PHATE, 
# and then putting PHATE embeddings back into seu object as a new reduction
# https://gist.github.com/jebard/cea2d161879203ca6e2aa05f32a031bf

# first pull out the cells we know to be involved in photoreceptor development
levels(organoid_all_tp)
organoid_prc_lineage <- subset(organoid_all_tp, subset = draft_celltype_v2 %in% c("Progenitors", 
                                                                                  "T1", "T3",
                                                                                  "Rod", 
                                                                                  "Cone"))

# run SCTransform...
organoid_prc_lineage <- SCTransform(organoid_prc_lineage, verbose = TRUE)

# pull out the SCT data
seurat_data <- GetAssayData(organoid_prc_lineage, assay = "SCT")
organoid_prc_lineage@assays$SCT@data

## reshape for input into PHATE
phate_data_input <- t(seurat_data)

# run PHATE
phate_outputv2 <- phate(phate_data_input, knn=6, decay=50, t=100)


## quick sanity check of the phate embeddings
ggplot(phate_outputv2) +
  geom_point(aes(PHATE1, PHATE2))

## stash the embeddings back into the seurat object as a dimension reduction object
organoid_prc_lineage[["PHATE"]] <- CreateDimReducObject(embeddings = phate_outputv2$embedding, key = "PHATE_", assay = DefaultAssay(organoid_prc_lineage))

# plot using seurat's default tools
p1 <- DimPlot(organoid_prc_lineage, reduction = "PHATE", group.by = "timepoint", label = T, repel = T, split.by = "line") + ggtitle(label = "PHATE") + NoLegend() + xlim (c(-0.03, 0.05)) + ylim (c(-0.015, 0.015))

p2 <- DimPlot(organoid_prc_lineage, reduction = "PHATE", group.by = "draft_celltype_v2", label = T, repel = T, split.by = "line") + ggtitle(label = "PHATE") + NoLegend() + xlim (c(-0.03, 0.05)) + ylim (c(-0.015, 0.015))

p3 <- DimPlot(organoid_prc_lineage, reduction = "PHATE", group.by = "line", label = T, repel = T, split.by = "draft_celltype_v2") + ggtitle(label = "PHATE") + NoLegend() + xlim (c(-0.03, 0.05)) + ylim (c(-0.015, 0.015))

p4 <- DimPlot(organoid_prc_lineage, reduction = "PHATE", split.by =  "line", label = T, repel = T) + ggtitle(label = "PHATE") + NoLegend() + xlim (c(-0.03, 0.05)) + ylim (c(-0.015, 0.015))

p5 <- DimPlot(organoid_prc_lineage, reduction = "PHATE", group.by = "draft_celltype_v2", label = F) + ggtitle(label = "PHATE") + xlim (c(-0.03, 0.05)) + ylim (c(-0.015, 0.015))
p1
p2
p3
p4
p5


pdf(file = "PHATE_umap_clusters.pdf", width = 10, height = 10)
p5
dev.off()

pdf(file = "PHATEv2_umap_clusters.pdf", width = 10, height = 10)
p5
dev.off()
```

############################################
###### CLUSTERING ON PHATE EMBEDDINGS ######
############################################

```{r, fig.width=30, fig.height=10}
organoid_prc_lineage <- FindNeighbors(organoid_prc_lineage, dims = 1:2, reduction = "PHATE")
organoid_prc_lineage <- FindClusters(organoid_prc_lineage, resolution = 0.5, algorithm = 3)

tiff("PHATE_clusters_split_lines.tiff", units="in", width=15, height=5, res=300)
DimPlot(organoid_prc_lineage, reduction = "PHATE", split.by = "line", label = T, group.by = "SCT_snn_res.0.5")
dev.off()

tiff("PHATE_clusters_split_timepoint.tiff", units="in", width=15, height=5, res=300)
DimPlot(organoid_prc_lineage, reduction = "PHATE", split.by = "timepoint", label = T, group.by = "SCT_snn_res.0.5")
dev.off()

tiff("UMAP_clusters_split_timepoint.tiff", units="in", width=15, height=5, res=300)
DimPlot(organoid_prc_lineage, group.by = "draft_celltype_v2", reduction = "PHATE", split.by = "timepoint", label = T)
dev.off()


# re-classify PHATE clusters using previous annotations.
table(organoid_prc_lineage$SCT_snn_res.0.5, organoid_prc_lineage$draft_celltype_v2)
table(organoid_prc_lineage$SCT_snn_res.0.5, organoid_prc_lineage$line)

Idents(organoid_prc_lineage) <- "SCT_snn_res.0.5"
levels(organoid_prc_lineage)
DimPlot(organoid_prc_lineage, reduction = "PHATE", label = T)

new.cluster.ids <- c("eCone" ,"DivRod" , "DivRod" , "imCone" , "eCone" , "Cone" , "eRod" , "eCone" , "Rod"  ,"Rod" ,"T3" ,
"Cone" ,"Rod" ,"eRod" ,"Rod", "T1", "imRod", "Rod", "Progenitors", "imRod", "Progenitors", 
"imRod", "Cone" ,"Cone", "Rod", "Progenitors" ,"DivRod" ,"DivRod" ,"Progenitors", "Cone", "imRod",
 "T3", "T3", "Progenitors", "Cone","Progenitors", "T1", "Rod", "Cone", "Progenitors", "Progenitors",
  "Progenitors", "Rod" ,"T1" ,"T1" ,"Cone")

names(new.cluster.ids) <- levels(organoid_prc_lineage)

organoid_prc_lineage <- RenameIdents(organoid_prc_lineage, new.cluster.ids)
organoid_prc_lineage$celltype_PHATEann <-  Idents(organoid_prc_lineage)

my_cols <- c(brewer.pal(n = 10, name = "Paired"), brewer.pal(n = 10, name = "Dark2"))
show_col(my_cols)

my_phate_cols <- " "

levels(organoid_prc_lineage)

DimPlot(object = organoid_prc_lineage, reduction = "PHATE", group.by = "celltype_PHATEann", label = T)
```



```{r}
Idents(organoid_prc_lineage) <- "celltype_PHATEann"

pdf("celltype_PHATE_clusters.pdf", width=10, height=10)
DimPlot(organoid_prc_lineage, label = T, group.by = "celltype_PHATEann", pt.size = 0.1, reduction = "PHATE") + NoLegend()
dev.off()

pdf("celltype_umap_prc_lin_clusters.pdf", width=10, height=10)
DimPlot(organoid_prc_lineage, label = T, group.by = "draft_celltype_v2", pt.size = 0.1, reduction = "umap") + NoLegend()
dev.off()

pdf("celltype_PHATE_clusters_split_by_line.pdf", width=25, height=10)
DimPlot(organoid_prc_lineage, label = T, group.by = "celltype_PHATEann", split.by = "line",pt.size = 0.1, reduction = "PHATE") + NoLegend()
dev.off()

pdf("celltype_PHATE_clusters_in_UMAP.pdf", width=10, height=10)
DimPlot(organoid_prc_lineage, label = T, group.by = "celltype_PHATEann",pt.size = 0.1, reduction = "umap") + NoLegend()
dev.off()

pdf("dimplot_celltype_PHATE_clusters_in_UMAP_by_time.pdf", width=20, height=5)
DimPlot(organoid_prc_lineage, label = T, group.by = "celltype_PHATEann",pt.size = 0.1, split.by = "timepoint", reduction = "umap") + NoLegend()
dev.off()

save(organoid_prc_lineage, file = "organoid_prc_lineage_only_PHATEann.RData")

```


# Marker genes of clusters
```{r}
#fix order 
levels(organoid_prc_lineage$celltype_PHATEann)

organoid_prc_lineage$celltype_PHATEann <- factor(x = organoid_prc_lineage$celltype_PHATEann, levels = c("Progenitors", 
                                                                                                              "T1",
                                                                                                              "T3",
                                                                                                              "eCone", "imCone", "Cone"    ,  
                                                                                                              "eRod", "imRod", "Rod",    "DivRod"))

Idents(organoid_prc_lineage) <- "celltype_PHATEann"

prc_lineage_markers <- FindAllMarkers(organoid_prc_lineage, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.5, assay = "RNA")

prc_lineage_markers %>%
    group_by(cluster) %>%
    top_n(n = 10, wt = avg_log2FC) -> top10


organoid_prc_lineage.small <- subset(organoid_prc_lineage, downsample = 500)

mapal <- colorRampPalette(RColorBrewer::brewer.pal(9,"RdBu"))(256)
pdf("prc_lineage_heatmap.pdf", width = 15, height = 10)
DoHeatmap(object = organoid_prc_lineage.small, features = top10$gene, group.by = "celltype_PHATEann", assay = "RNA", raster = F) + NoLegend()+
  scale_fill_gradientn(colours = rev(mapal))
dev.off()

prc_lineage_markers <- c("TOP2A", "ATOH7", "PHLDA1", "RCVRN", "GNAT1", "GNGT1", "RHO", "ARR3", "PDE6H", "NRL", "CRX", "OTX2", "MKI67")

tiff("prc_lineage_dotplot.tiff", units="in", width=8, height=5, res=300)
DotPlot(object = organoid_prc_lineage, features = prc_lineage_markers, group.by = "celltype_PHATEann", assay = "RNA") + theme(axis.text.x = element_text(angle = 90)) + scale_colour_gradient2(low = "#FFF5F0", mid = "#fb6a4a", high = "#67000d")
dev.off()

```

########################################
###### SLINGSHOT ON SEURAT OBJECT ######
########################################
```{r}
cell_pal <- function(cell_vars, pal_fun,...) {
  if (is.numeric(cell_vars)) {
    pal <- pal_fun(100, ...)
    return(pal[cut(cell_vars, breaks = 100)])
  } else {
    categories <- sort(unique(cell_vars))
    pal <- setNames(pal_fun(length(categories), ...), categories)
    return(pal[cell_vars])
  }
}

cell_colors <- cell_pal(organoid_prc_lineage@meta.data$celltype_PHATEann, brewer_pal(12, "Paired"))
cell_colors_clust <- cell_pal(organoid_prc_lineage@meta.data$celltype_PHATEann, hue_pal())


sds_phate <- NULL
sds_phate <- slingshot(Embeddings(organoid_prc_lineage, "PHATE"), 
                       clusterLabels = organoid_prc_lineage$celltype_PHATEann, extend = "n",                       
                       start.clus = c("Progenitors"),
                       end.clus = c("Cone", "DivRod", "Rod"), stretch = 0.1, thresh = 0.01,  approx_points = 150)


sds <- SlingshotDataSet(sds_phate)

save(sds, file = "sds_phate_prc_lineage.RData")

pdf("PHATE_prc_slingshot_lineages.pdf", width=10, height=10)
plot(reducedDim(sds), col = cell_colors, pch = 16, cex = 0.5)
lines(sds, lwd = 2, col = 'black', type = 'lineages')
dev.off()

pdf("PHATE_prc_slingshot_curves.pdf", width=10, height=10)
plot(reducedDim(sds), col = cell_colors, pch = 16, cex = 0.5)
lines(sds, lwd = 2, col = 'black', type = 'curves')
dev.off()

nc <- 3
pt <- slingPseudotime(sds)
pt
nms <- colnames(pt)
nms <- nms[1:3]
nms
nr <- ceiling(length(nms)/nc)
pal <- viridis(100, end = 0.95)

pdf("PHATE_prc_slingshot_lineage_pseudotime.pdf", width=30, height=10)
par(mfrow = c(nr, nc))
for (i in nms) {
  colors <- pal[cut(pt[,i], breaks = 100)]
  plot(reducedDim(sds), col = colors, pch = 16, cex = 0.5, main = i)
  lines(sds, lwd = 4, col = 'black', type = 'lineages')
}


colors <- pal[cut(pt[,i], breaks = 100)]

paletteer_c(viridis)
image(viridis(100, end = 0.95))

paletteer_c("ggthemes::Classic Area Green", 30)
ddf<-viridis(100, end = 0.95)

image(1:nrow(ddf), 1, as.matrix(1:nrow(ddf)), 
      col=rgb(ddf$r, ddf$g, ddf$b),
      xlab="", ylab = "", xaxt = "n", yaxt = "n", bty = "n")
viridis(100, end = 0.95)

install.packages("paletteer")
viridis(100, end = 0.95)
dev.off()

```

## TRADESEQ ##

```{r}
#https://github.com/NBISweden/single-cell_sib_scilifelab/blob/master/session-trajectories/3_slingshot.md

BiocParallel::register(BiocParallel::SerialParam())

# filter - not sure which counts to use...
organoid_prc_lineage <- FindVariableFeatures(organoid_prc_lineage, selection.method = "vst",assay = "RNA", nfeatures = 2000) # use 2000 to speed up fitGAM... hopefully
var_feats <- organoid_prc_lineage@assays$RNA@var.features
counts <- organoid_prc_lineage@assays$RNA@counts
filt <- row.names(counts) %in% var_feats
counts <- counts[filt, ]


#icMat <- evaluateK(counts = counts, sds = sds_phate, k = 3:7, 
                   #nGenes = 100, verbose = T)


sce <- fitGAM(
  counts = as.matrix(counts),
  sds = sds,
  parallel = T
)

save(sce, file = "fitGAM_obj.RData")

```



pseudotime plots of genes

```{r}


plotGeneCount(curve = sds, counts = organoid_prc_lineage@assays$RNA@counts, gene = "PDE6H")


pdf(file = "PDE6H_smoothers.pdf", width = 8, height = 5)
plotSmoothers(models = sce, counts = organoid_prc_lineage@assays$RNA@counts, gene = "PDE6H", curvesCols = c("#F69799", "#E21F26", "#CAC5E2"), alpha = 0.05) +
  scale_color_manual(values = c("#F69799", "#E21F26", "#CAC5E2"))
dev.off()


pdf(file = "NRL_smoothers.pdf", width = 8, height = 5)
plotSmoothers(models = sce, counts = organoid_prc_lineage@assays$RNA@counts, gene = "NRL", curvesCols = c("#F69799", "#E21F26", "#CAC5E2")) + scale_color_manual(values = c("#F69799", "#E21F26", "#CAC5E2"))
dev.off()

pdf(file = "NR2E3_smoothers.pdf", width = 8, height = 5)
plotSmoothers(models = sce, counts = organoid_prc_lineage@assays$RNA@counts, gene = "NR2E3", curvesCols = c("#F69799", "#E21F26", "#CAC5E2")) + scale_color_manual(values = c("#F69799", "#E21F26", "#CAC5E2"))
dev.off()

pdf(file = "RHO_smoothers.pdf", width = 8, height = 5)
plotSmoothers(models = sce, counts = organoid_prc_lineage@assays$RNA@counts, gene = "RHO", curvesCols = c("#F69799", "#E21F26", "#CAC5E2")) + scale_color_manual(values = c("#F69799", "#E21F26", "#CAC5E2"))
dev.off()

pdf(file = "GNAT1_smoothers.pdf", width = 8, height = 5)
plotSmoothers(models = sce, counts = organoid_prc_lineage@assays$RNA@counts, gene = "GNAT1", curvesCols = c("#F69799", "#E21F26", "#CAC5E2")) + scale_color_manual(values = c("#F69799", "#E21F26", "#CAC5E2"))
dev.off()
```

## PLOTTING PSEUDOTIME CURVES TO SUPPORT NORMAL DIFF FIG 2
```{r}
 load("~/LSS/IVR/nkmullin/experiments/sc-rna-seq/nr2e3_ips_ro_timecourse/seurat_downstream/phate_pseudotime_trajectory/fitGAM_obj.RData")

names <- list(
  "1"="Cone",
  "2"="Rod",
  "3"="Divergent Rod"
)

prc_labeller <- function(variable,value){
  return(names[value])
}


SplitPseudoPlotter <- function(gene_symbol) { # create a function with the name my_function

  plot <- plotSmoothers(models = sce, counts = organoid_prc_lineage@assays$RNA@counts, gene = gene_symbol, curvesCols = c("#F69799", "#E21F26", "#CAC5E2")) + 
   scale_color_manual(values = c("#F69799", "#E21F26", "#CAC5E2")) + NoLegend() + ylab(label = paste0(gene_symbol)) +
    facet_wrap(~lineage,  labeller = prc_labeller)

  pdf(file = paste0(gene_symbol,"_split_smoothers.pdf"), width = 6, height = 2)
    print(plot)
  dev.off()

}


SplitPseudoPlotter <- function(gene_symbol_vec, name) { # create a function with the name my_function

  num_genes <- length(gene_symbol_vec)
  plot_list <- list()
  SplitPseudoPlotter_Sub <- function(gene_symbol) { # create a function with the name my_function
    plot <- plotSmoothers(models = sce, counts = organoid_prc_lineage@assays$RNA@counts, gene = gene_symbol, curvesCols = c("#F69799", "#E21F26", "#CAC5E2")) + 
      scale_color_manual(values = c("#F69799", "#E21F26", "#CAC5E2")) + NoLegend() + ylab(label = paste0(gene_symbol)) +
        facet_wrap(~lineage,  labeller = prc_labeller)
 
   return(plot)
    }
  
  
  plot_list <- lapply(gene_symbol_vec,SplitPseudoPlotter_Sub)

  pdf(file = paste0(name,"_split_smoothers.pdf"), width = 6, height = 2*num_genes)
    print(ggarrange(plotlist = plot_list, ncol = 1, labels = "AUTO"))
  dev.off()

}

###
prog_specific <- c("PAX6", "LHX2", "HES1")
SplitPseudoPlotter(prog_specific, "prog_specific")

T1_specific <- c("GADD45A", "ATOH7", "RGS16", "GAL")
SplitPseudoPlotter(T1_specific, "T1_specific")

T3_specific <- c("FABP7", "DLL3")
SplitPseudoPlotter(T3_specific, "T3_specific")

PR_specific <- c("PRDM1", "RCVRN")
SplitPseudoPlotter(PR_specific, "PR_specific")

rod_specific <- c("NRL", "NR2E3", "GNAT1")
SplitPseudoPlotter(rod_specific, "rod_specific")

cone_specific <- c("THRB", "OPN1SW", "PDE6H")
SplitPseudoPlotter(cone_specific, "cone_specific")

synapse_prog <- c("NEUROG1", "SYNDIG1")
SplitPseudoPlotter(synapse_prog, "synapse_prog")
```

