---
title: "add_motifs_run_chromvar"
author: "Nate Mullin"
date: "03/28/2023"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(Signac)
library(Seurat)
library(patchwork)
library(tidyverse)
library(EnsDb.Hsapiens.v86)
library(Matrix)
library(SummarizedExperiment)
library(ggpubr)
library(cowplot)
library(JASPAR2022)
library(TFBSTools)
library(BSgenome.Hsapiens.UCSC.hg38)
library(patchwork)
library(EnhancedVolcano)
set.seed(1234)
```


```
load data (annotated multiome data without CHROMVAR or MOTIF assay yet.)
```{r}
load(".../wnn_analysis/multimodal_object_02382023.RData")
DefaultAssay(multimodal_object) <- 'ATAC'
```

#################################################################################
############################ Recall Peaks with MACS2 ############################
#################################################################################
```{r}
peaks <- CallPeaks(
  object = multimodal_object,
  group.by = "celltype" , macs2.path = "~/.local/bin/macs2"
  )

# remove peaks on nonstandard chromosomes and in genomic blacklist regions
peaks <- keepStandardChromosomes(peaks, pruning.mode = "coarse")
peaks <- subsetByOverlaps(x = peaks, ranges = blacklist_hg38_unified, invert = TRUE)

# quantify counts in each peak
macs2_counts <- FeatureMatrix(
  fragments = Fragments(multimodal_object),
  features = peaks,
  cells = colnames(multimodal_object)
)

# create a new assay using the MACS2 peak set and add it to the Seurat object
multimodal_object[["macs2_peaks"]] <- CreateChromatinAssay(
  counts = macs2_counts,
  fragments = multimodal_object@assays$ATAC@fragments,
  annotation = multimodal_object@assays$ATAC@annotation
)

save(multimodal_object, file="multimodal_object_macs2_peaks_03282023.RData")
```

