---
title: "add_motifs_run_chromvar"
author: "Nate Mullin"
date: "03/28/2023"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(Signac)
library(Seurat)
library(patchwork)
library(tidyverse)
library(EnsDb.Hsapiens.v86)
library(Matrix)
library(SummarizedExperiment)
library(ggpubr)
library(cowplot)
library(JASPAR2022)
library(TFBSTools)
library(BSgenome.Hsapiens.UCSC.hg38)
library(patchwork)
library(EnhancedVolcano)
set.seed(1234)
```


```
load data (annotated multiome data without CHROMVAR or MOTIF assay yet.)
```{r}
load("~/LSS/IVR/nkmullin/experiments/multiome/XXXXX")
DefaultAssay(multimodal_object) <- 'ATAC'
```

#################################################################################
############################ Recall Peaks with MACS2 ############################
#################################################################################
```{r}
peaks <- CallPeaks(
  object = multimodal_object,
  group.by = "celltype" , macs2.path = "~/.local/bin/macs2"
  )

# remove peaks on nonstandard chromosomes and in genomic blacklist regions
peaks <- keepStandardChromosomes(peaks, pruning.mode = "coarse")
peaks <- subsetByOverlaps(x = peaks, ranges = blacklist_hg38_unified, invert = TRUE)

# quantify counts in each peak
macs2_counts <- FeatureMatrix(
  fragments = Fragments(multimodal_object),
  features = peaks,
  cells = colnames(multimodal_object)
)

# create a new assay using the MACS2 peak set and add it to the Seurat object
multimodal_object[["macs2_peaks"]] <- CreateChromatinAssay(
  counts = macs2_counts,
  fragments = multimodal_object@assays$ATAC@fragments,
  annotation = multimodal_object@assays$ATAC@annotation
)

save(multimodal_object, file="multimodal_object_macs2_peaks_03282023.RData")
```

```{r}
load("~/LSS/IVR/nkmullin/experiments/multiome/nr2e3_d160_d280_organoid/seurat_downstream/motif/multimodal_object_macs2_peaks_03282023.RData")
```


###############################################################################
############################ motif enrichment DARs ############################
###############################################################################
```{r}
# set idents/assay
Idents(multimodal_object) <- "manual_celltype_rough"
DefaultAssay(multimodal_object) <- "macs2_peaks"

# calculate DA peaks in Rods
# Positive = Closed in NR2E3-null, open in CTL (Normally open in Control) - putative NRL etc sites
# Negative = Open in B342, (Normally closed in Control) - putative repressed sites

# find DA peaks in ctl vs 342 in ROD only
rod_da_peaks <- FindMarkers(
  object = multimodal_object, group.by = "line", subset.ident = c("Rod"),
  ident.1 = 'B342cor',
  ident.2 = 'B342',
  only.pos = FALSE,
  test.use = 'LR',
  latent.vars = 'nCount_macs2_peaks'
)

write.csv(rod_da_peaks, file = "rod_da_peaks_B342cor_vs_B342.csv")

rod_da_peaks <- read_csv("rod_da_peaks_B342cor_vs_B342.csv")
rod_da_peaks <- column_to_rownames(rod_da_peaks, var = "...1")

pdf(file = "rod_da_peaks_B342cor_vs_B342.pdf", width = 10, height = 10)
EnhancedVolcano(toptable = rod_da_peaks, 
                x = "avg_log2FC", y = "p_val_adj", 
                lab = rownames(rod_da_peaks), pCutoff = 0.05, FCcutoff = 1, drawConnectors = T)
dev.off()
```

Plot DARs

```{r}
obj <- rod_da_peaks
obj$delta.pct <- obj$pct.1 - obj$pct.2
obj$peak <- rownames(obj)

dpct_fc_plot <- ggplot(obj, aes(delta.pct, avg_log2FC)) + 
  geom_point(color = "#bdbdbd", alpha=I(0.01)) +
  geom_point(data=subset(obj, pct.1 > 0.1 & avg_log2FC > 0 & p_val_adj < 0.05),
             aes(delta.pct,avg_log2FC,label = peak), color = "#b2182b",alpha=I(0.5)) + 
  geom_point(data=subset(obj, pct.2 > 0.1 & avg_log2FC < 0 & p_val_adj < 0.05),
             aes(delta.pct,avg_log2FC,label = peak), color  = "#2166ac",alpha=I(0.5)) + 
  labs(x = "âˆ† Percent Expressed", y = "Log2 Fold Change", title = "Rod") +
  theme(plot.title = element_text(hjust = 0.5)) + theme_bw() 

dpct_fc_plot

tiff(file = "rod_dar_used_in_motif_adjp05_pct0.1.tiff", width = 5, height = 5, res = 300, units = "in")
dpct_fc_plot
dev.off()

```


Pull up and down rod peaks into two lists
```{r}
sig_adj_rod_da_peaks <- subset(rod_da_peaks, p_val_adj < 0.05)
sig_adj_rod_da_peaks <- subset(sig_adj_rod_da_peaks, pct.1 > 0.1 |pct.2 > 0.1 )

pos_peaks <- rownames(sig_adj_rod_da_peaks[sig_adj_rod_da_peaks$avg_log2FC > 0,])
neg_peaks <- rownames(sig_adj_rod_da_peaks[sig_adj_rod_da_peaks$avg_log2FC < 0,])

# match the overall GC content in the peak set
multimodal_object <- RegionStats(multimodal_object, genome = BSgenome.Hsapiens.UCSC.hg38)
open.peaks <- AccessiblePeaks(multimodal_object, idents = c("Rod"))
meta.feature <- GetAssayData(multimodal_object, assay = "macs2_peaks", slot = "meta.features")

pos.peaks.matched <- MatchRegionStats(
  meta.feature = meta.feature[open.peaks, ],
  query.feature = meta.feature[pos_peaks, ],
  n = 50000
)

open_b342cor_motif <- FindMotifs(
  object = multimodal_object,
  features = pos_peaks, 
  background = pos.peaks.matched
)

neg.peaks.matched <- MatchRegionStats(
  meta.feature = meta.feature[open.peaks, ],
  query.feature = meta.feature[neg_peaks, ],
  n = 50000
)

open_b342_motif <- FindMotifs(
  object = multimodal_object,
  features = neg_peaks, 
  background = neg.peaks.matched
)

open_b342cor_motif
open_b342_motif
```

Plot
```{r}
library(ggrepel)
# plot adding up all layers we have seen so far
rod_object <- subset(multimodal_object, manual_celltype_rough == "Rod")


rod_sum_greater_1000 <- function(gene) {
  DefaultAssay(rod_object) <- "RNA"
  return(tryCatch(FetchData(object = rod_object, vars = gene, slot = "counts") %>% sum() %>% ifelse(.>1000), error=function(e) NULL))
}

rod_sum_greater_1000

filt <- open_b342_motif$motif.name %in% rownames(rod_object@assays$RNA)
open_b342_motif_filt <- open_b342_motif[filt,]


results <- list()
open_b342_motif_filt$motif.name
for (x in open_b342_motif_filt$motif.name) {
  val <- rod_sum_greater_1000(x)
  # results[x] <- val
} 

results

open_b342_motif_mut <- mutate(open_b342_motif, rod_sum_greater_1000_list(motif.name)) 
rod_sum_greater_1000_list(toupper(open_b342_motif$motif.name))

toupper(open_b342_motif$motif.name)

open_b342_motif[1:10] %>%
  mutate(exp_1000 = rod_sum_greater_1000(motif.name))


highlight <- dplyr::filter(open_b342_motif, motif.name %in% c("NRL", "Nr2e3"))
highlight_up <- dplyr::filter(open_b342_motif, fold.enrichment > 1.5 & -log10(p.adjust) > 10)

open_b342_motif_plot <- ggplot(data=open_b342_motif, aes(x=fold.enrichment, y=-log10(p.adjust), label=motif.name)) +
        geom_point() + 
        theme_minimal() +
        geom_label_repel(data = highlight_up, box.padding = 0.5, max.overlaps = 30, fill = "gray") +
        scale_color_manual(values=c("blue", "black", "red")) +
        geom_point(data=highlight, 
             aes(x=fold.enrichment,y=-log10(p.adjust)), 
             color='red') +
        geom_label_repel(data = highlight, box.padding = 0.5, max.overlaps = 10, fill = "white") +
  labs(title = "Enriched Motifs in OPEN peaks in B342", x = "Fold Enrichment over Background")

highlight <- dplyr::filter(open_b342cor_motif, motif.name %in% c("NRL", "Nr2e3"))
highlight_up <- dplyr::filter(open_b342cor_motif, fold.enrichment > 1.5 & -log10(p.adjust) > 2)

open_b342cor_motif_plot <- ggplot(data=open_b342cor_motif, 
                                  aes(x=fold.enrichment, y=-log10(p.adjust), label=motif.name)) +
        geom_point() + 
        theme_minimal() +
        geom_label_repel(data = highlight_up, box.padding = 0.5, max.overlaps = 30, fill = "gray") +
        scale_color_manual(values=c("blue", "black", "red")) +
        geom_point(data=highlight, 
             aes(x=fold.enrichment,y=-log10(p.adjust)), 
             color='red') +
        geom_label_repel(data = highlight, box.padding = 0.5, max.overlaps = 10, fill = "white") +
  labs(title = "Enriched Motifs in CLOSED Peaks in B342", x = "Fold Enrichment over Background") 

open_b342cor_motif_plot

pdf(file = "b342_motif_open_closed.pdf", width = 10, height = 5)
ggarrange(open_b342_motif_plot,open_b342cor_motif_plot, ncol = 2, nrow = 1)
dev.off()
```


FOOTPRINT
```{r}
# gather the footprinting information for sets of motifs
multimodal_object <- Footprint(
  object = multimodal_object,
  motif.name = c("NRL", "Nr2e3"),
  genome = BSgenome.Hsapiens.UCSC.hg38
)
rod_only <- subset(multimodal_object, manual_celltype_rough == "Rod")

rod_only <- Footprint(
  object = rod_only,
  motif.name = c("NRL", "Nr2e3"),
  genome = BSgenome.Hsapiens.UCSC.hg38
)
# plot the footprint data for each group of cells
Idents(multimodal_object) <- "manual_celltype_rough"
nrl_fp <- PlotFootprint(multimodal_object, features = c("NRL"), split.by = "manual_celltype_rough", show.expected = F, group.by = "line", label = F)
nr2e3_fp <- PlotFootprint(multimodal_object, features = c("Nr2e3"), split.by = "manual_celltype_rough", show.expected = F,  group.by = "line", label = F)


nrl_fp

pdf(file = "footprint_nrl_nr2e3.pdf", width = 10, height = 16)
ggarrange(nrl_fp, nr2e3_fp, ncol = 1, nrow = 2)
dev.off()

```

